src/core_api.cpp

```diff
@@ -1693,3 +1695,244 @@ bool del_preset(const std::shared_ptr<http_req>& req, const std::shared_ptr<http
+        }
+    }
+
+    if(new_rule.action == RateLimitAction::throttle)
+    {
+        if(req_json.count("max_requests_60s") == 0 || req_json.count("max_requests_1h") == 0) {
+            res->set_400("Both `max_requests_60s` and `max_requests_1h` are required for throttling");
+            return false;
+        }
+
+        if(req_json["max_requests_60s"].get<int>() < 1 || req_json["max_requests_1h"].get<int>() < 1) {
+            res->set_400("Both `max_requests_60s` and `max_requests_1h` must be greater than 0");
+            return false;
+        }
+
+        new_rule.max_requests.minute_threshold = req_json["max_requests_60s"];
+        new_rule.max_requests.hour_threshold = req_json["max_requests_1h"];
+
+        if((req_json.count("auto_ban_threshold_num") > 0 && req_json.count("auto_ban_num_days") == 0) || (req_json.count("auto_ban_threshold_num") == 0 && req_json.count("auto_ban_num_days") > 0)) {
+            res->set_400("Both `auto_ban_threshold_num` and `auto_ban_num_days` are required if either is specified");
+            return false;
+        }
+
+        if(req_json.count("auto_ban_threshold_num") > 0 && req_json.count("auto_ban_num_days") > 0) {
+
+            if(req_json["auto_ban_threshold_num"].get<int>() < 1 || req_json["auto_ban_num_days"].get<int>() < 1) {
+                res->set_400("Both `auto_ban_threshold_num` and `auto_ban_num_days` must be greater than 0");
+                return false;
+            }
+
+            new_rule.auto_ban_threshold_num = req_json["auto_ban_threshold_num"];
+            new_rule.auto_ban_num_days = req_json["auto_ban_num_days"];
+        }
+    }
+
+    if(rateLimitManager->edit_rule_by_id(std::stoull(req->params["id"]), new_rule)) {
+        res->set_200("OK");
+        return true;
+    } else {
+        res->set_400("Rule for one or more entities already exist");
+        return false;
+    }
+
+    return true;
+}
+
+bool del_rate_limit(const std::shared_ptr<http_req>& req, const std::shared_ptr<http_res>& res) {
+    RateLimitManager* rateLimitManager = RateLimitManager::getInstance();
+    uint64_t id = std::stoull(req->params["id"]);
+    const auto& ruleOption = rateLimitManager->find_rule_by_id(id);
+
+    if(!ruleOption.ok()) {
+        res->set_404();
+        return false;
+    }
+
+    rateLimitManager->delete_rule_by_id(id);
+    res->set_200("OK");
+    return true;
+}
+
+bool post_rate_limit(const std::shared_ptr<http_req>& req, const std::shared_ptr<http_res>& res) {
+    RateLimitManager* rateLimitManager = RateLimitManager::getInstance();
+    nlohmann::json req_json;
+
+    try {
+        req_json = nlohmann::json::parse(req->body);
+    } catch (const std::exception & e) {
+        res->set_400("Bad JSON.");
+        return false;
+    }
+
+    if (req_json.count("action") == 0) {
+        res->set_400("Parameter `action` is required.");
+        return false;
+    }
+
+    if (req_json["action"] == "allow") {
```

Comment:
All the processing below is duplicated with the `put_rate_limit` above. We should probably extract them out and maybe make them as part of rateLimitManager class.